<a href="https://colab.research.google.com/github/alexandraeremia06/Kaggle/blob/main/Statistical_testing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from google.colab import files
uploaded =files.upload()

Saving olist_order_payments_dataset.csv to olist_order_payments_dataset (1).csv
Saving olist_order_reviews_dataset.csv to olist_order_reviews_dataset (1).csv
Saving olist_products_dataset.csv to olist_products_dataset (1).csv
Saving olist_customers_dataset.csv to olist_customers_dataset (1).csv
Saving olist_order_items_dataset.csv to olist_order_items_dataset (1).csv
Saving olist_orders_dataset.csv to olist_orders_dataset (1).csv
Saving olist_sellers_dataset.csv to olist_sellers_dataset (1).csv


In [3]:
import pandas as pd
import numpy as np
import plotly.express as px
import datetime

In [4]:
customers = pd.read_csv('olist_customers_dataset.csv')
orders_items = pd.read_csv('olist_order_items_dataset.csv')
order_payments = pd.read_csv('olist_order_payments_dataset.csv')
order_reviews = pd.read_csv('olist_order_reviews_dataset.csv')
orders = pd.read_csv('olist_orders_dataset.csv')
products = pd.read_csv('olist_products_dataset.csv')
sellers = pd.read_csv('olist_sellers_dataset.csv')

In [5]:
print(customers.head())

                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  


In [6]:
print(orders_items.head())

                           order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.7

In [7]:
print(order_payments.head())

                           order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45  


In [8]:
print(order_reviews.head())

                          review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3           

In [9]:
print(orders.head())

                           order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08

In [10]:
print(products.head())

                         product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0       

In [11]:
print(sellers.head())

                          seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP  


In [12]:
#join tables orders + orders_review
df_combined = orders.merge(order_reviews, on='order_id', how='left')

In [13]:
df_combined.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  
0     2017-10-12 03:43:48  
1     2018-08-08 18:37:50  
2     2018-08-22 19:07:58  
3     2017-12-05 19:21:58  
4     2018-02-18 13:02:51

In [15]:
#transfom date to datetime
df_combined['order_purchase_timestamp'] = pd.to_datetime(df_combined['order_purchase_timestamp'])
df_combined['order_delivered_customer_date'] = pd.to_datetime(df_combined['order_delivered_customer_date'])

In [16]:
df_combined.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  
0     2017-10-12 03:43:48  
1     2018-08-08 18:37:50  
2     2018-08-22 19:07:58  
3     2017-12-05 19:21:58  
4     2018-02-18 13:02:51

In [17]:
df_combined['order_purchase_timestamp'].dtype

dtype('<M8[ns]')

In [18]:
df_combined['order_delivered_customer_date'].dtype

dtype('<M8[ns]')


STEP 2
**Shipping Time Statistics**

Mean shipping time: This gives the average time taken for orders to be delivered to customers. D days

Median shipping time: Similar to the median for order price, this represents the middle value of shipping times. E days

Standard deviation of shipping times: This measures the variability in shipping times. F days

Boxplot of shipping times: This visualizes the spread of shipping times, showing the minimum, maximum, median, and outliers.



In [19]:
#Mean shipping time
#new column with dif between delivered date and the purchase date
#avg of the new column
df_combined['shipping_time_days'] = (df_combined['order_delivered_customer_date'] - df_combined['order_purchase_timestamp']).dt.days
df_combined.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  shipping_time_days  
0     2017-10-12 03:43:48                 8.0  
1     2018-08-08 18:37:50                13.0  
2     2018-08-22 19:07:58                 9.0  
3     2017-12-05 19:21:58                13.0  
4     2018-02-18 13:02:51                 2.0

In [20]:
#selecting only the delivered from order status
df_delivered = df_combined[df_combined['order_status']== 'delivered']
df_delivered.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  shipping_time_days  
0     2017-10-12 03:43:48                 8.0  
1     2018-08-08 18:37:50                13.0  
2     2018-08-22 19:07:58                 9.0  
3     2017-12-05 19:21:58                13.0  
4     2018-02-18 13:02:51                 2.0

In [53]:
df_delivered['avg_shipping_time_days'] = df_delivered['shipping_time_days'].mean()
df_delivered.head()

#Avg shipping time is 12 days

/tmp/ipython-input-4019452031.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  shipping_time_days  avg_shipping_time_days  \
0     2017-10-12 03:43:48                 8.0               12.099537   
1     2018-08-08 18:37:50                13.0               12.099537   
2     2018-08-22 19:07:58                 9.0               12.099537   
3     2017-12-05 19:21:58                13.0               12.099537   
4     2018-02-18 13:02:51                 2.0               12.099537   

   median_shipping_time_days  std_shipping_time_days  on_time  
0                       10.0                9.547594     True  
1                       10.0                9.547594     True  
2                       10.0                9.547594     True  
3                       10.0                9.547594     True  
4                       10.0                9.547594     True

In [52]:
#Median shipping time: Similar to the median for order price, this represents the middle value of shipping times. E days
df_delivered['median_shipping_time_days'] = df_delivered['shipping_time_days'].median()
df_delivered.head()

#Median shipping time is 10 days

/tmp/ipython-input-3399833798.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  shipping_time_days  avg_shipping_time_days  \
0     2017-10-12 03:43:48                 8.0               12.099537   
1     2018-08-08 18:37:50                13.0               12.099537   
2     2018-08-22 19:07:58                 9.0               12.099537   
3     2017-12-05 19:21:58                13.0               12.099537   
4     2018-02-18 13:02:51                 2.0               12.099537   

   median_shipping_time_days  std_shipping_time_days  on_time  
0                       10.0                9.547594     True  
1                       10.0                9.547594     True  
2                       10.0                9.547594     True  
3                       10.0                9.547594     True  
4                       10.0                9.547594     True

In [51]:
#Standard deviation of shipping times
df_delivered['std_shipping_time_days'] = df_delivered['shipping_time_days'].std()
df_delivered.head()

#std = 9 days

/tmp/ipython-input-1069491731.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  shipping_time_days  avg_shipping_time_days  \
0     2017-10-12 03:43:48                 8.0               12.099537   
1     2018-08-08 18:37:50                13.0               12.099537   
2     2018-08-22 19:07:58                 9.0               12.099537   
3     2017-12-05 19:21:58                13.0               12.099537   
4     2018-02-18 13:02:51                 2.0               12.099537   

   median_shipping_time_days  std_shipping_time_days  on_time  
0                       10.0                9.547594     True  
1                       10.0                9.547594     True  
2                       10.0                9.547594     True  
3                       10.0                9.547594     True  
4                       10.0                9.547594     True

In [50]:
#Boxplot of shipping times: This visualizes the spread of shipping times, showing the minimum, maximum, median, and outliers.

fig = px.box(
    df_delivered,
    x='shipping_time_days'
)
fig.show()

In [49]:
df_delivered.columns


Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'review_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp', 'shipping_time_days',
       'avg_shipping_time_days', 'median_shipping_time_days',
       'std_shipping_time_days', 'on_time'],
      dtype='object')

STEP 3
**Testing the Effect of Shipping Time on Customer Satisfaction**

Hypothesis 1:
Null Hypothesis (H0): There is no difference in customer satisfaction between orders delivered on time and orders delivered late.
Alternative Hypothesis (H1): Orders delivered on time have higher customer satisfaction compared to orders delivered late.

A/B Testing Process:
Divide the dataset into two groups: one group with orders delivered before or on the estimated delivery time and the other with orders delivered after the estimated delivery time.
Compute the average satisfaction rating for each group.
Conduct a hypothesis test (e.g., t-test) to determine if there is a statistically significant difference in satisfaction ratings between the two groups.

Interpretation:
If the p-value is less than a chosen significance level (e.g., 0.05), reject the null hypothesis and conclude that there is a significant difference in satisfaction ratings between on time and late deliveries.
If the p-value is greater than the significance level, fail to reject the null hypothesis, indicating no significant difference in satisfaction ratings between on time and late deliveries.

In [54]:
#check if the orders were delivered on time or not
# order_delivered_customer_date is bigger than order_estimated_delivery_date it was late
on_time = df_delivered[df_delivered['order_delivered_customer_date'] <= df_delivered['order_estimated_delivery_date']]
on_time

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99987  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99988  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99989  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99990  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99991  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99987    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99988    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99989    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99990    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99991    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99987          2017-03-10 11:18:03           2017-03-17 15:08:01   
99988          2018-02-07 23:22:42           2018-02-28 17:37:56   
99989          2017-08-28 20:52:26           2017-09-21 11:24:17   
99990          2018-01-12 15:35:03           2018-01-25 23:32:54   
99991          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date                         review_id  \
0               2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1               2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2               2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3               2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4               2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   
...                             ...                               ...   
99987           2017-03-28 00:00:00  e262b3f92d1ce917aa412a9406cf61a6   
99988           2018-03-02 00:00:00  29bb71b2760d0f876dfa178a76bc4734   
99989           2017-09-27 00:00:00  371579771219f6db2d830d50805977bb   
99990           2018-02-15 00:00:00  8ab6855b9fe9b812cd03a480a25058a1   
99991           2018-04-03 00:00:00  dc9c59b4688062c25758c2be4cafc523   

       review_score review_comment_title  \
0               4.0                  NaN   
1               4.0     Muito boa a loja   
2               5.0                  NaN   
3               5.0                  NaN   
4               5.0                  NaN   
...             ...                  ...   
99987           5.0                  NaN   
99988           4.0                  NaN   
99989           5.0                  NaN   
99990           2.0                  NaN   
99991           5.0                  NaN   

                                  review_comment_message review_cr

Hypothesis 1: Null Hypothesis (H0): There is no difference in customer satisfaction between orders delivered on time and orders delivered late. Alternative Hypothesis (H1): Orders delivered on time have higher customer satisfaction compared to orders delivered late.

In [55]:
late = df_delivered[df_delivered['order_delivered_customer_date'] > df_delivered['order_estimated_delivery_date']]

In [56]:
on_time = df_delivered[df_delivered['order_delivered_customer_date'] <= df_delivered['order_estimated_delivery_date']]
late = df_delivered[df_delivered['order_delivered_customer_date'] > df_delivered['order_estimated_delivery_date']]

In [57]:
# order_delivered_customer_date is bigger than order_estimated_delivery_date it was late - boolean column with result
df_delivered['on_time'] = (df_delivered['order_delivered_customer_date'] <= df_delivered['order_estimated_delivery_date'])
df_delivered.head()

/tmp/ipython-input-1852510561.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  shipping_time_days  avg_shipping_time_days  \
0     2017-10-12 03:43:48                 8.0               12.099537   
1     2018-08-08 18:37:50                13.0               12.099537   
2     2018-08-22 19:07:58                 9.0               12.099537   
3     2017-12-05 19:21:58                13.0               12.099537   
4     2018-02-18 13:02:51                 2.0               12.099537   

   median_shipping_time_days  std_shipping_time_days  on_time  
0                       10.0                9.547594     True  
1                       10.0                9.547594     True  
2                       10.0                9.547594     True  
3                       10.0                9.547594     True  
4                       10.0                9.547594     True

In [86]:
#verify assumptions

from scipy.stats import shapiro

on_time_scores = on_time['review_score'].dropna()
late_scores = late['review_score'].dropna()

shapiro_on_time = shapiro(on_time_scores)
shapiro_late = shapiro(late_scores)

#print("On-time:", shapiro_on_time)

#On-time: ShapiroResult(statistic=np.float64(0.6543380333977209), pvalue=np.float64(8.060000311605506e-146))

print("Late:", shapiro_late)

#Late: ShapiroResult(statistic=np.float64(0.7758994955348266), pvalue=np.float64(3.961329385975819e-73))
#p > 0.05 → distribution ~ normal
#p ≤ 0.05 → not normal

Late: ShapiroResult(statistic=np.float64(0.7759100405250555), pvalue=np.float64(4.187913153305576e-73))


In [83]:
#since we rejected normality we want to use a nonparametric test like mann whitneyu
from scipy.stats import mannwhitneyu
t_stat, p_value = mannwhitneyu(on_time_scores, late_scores,alternative='greater')
print(f"U statistic: {t_stat}")
print(f"p-value: {p_value}")

U statistic: 530208489.0
p-value: 0.0


In [99]:
from scipy.stats import ttest_ind

# t test
# We compare the 'review_score' of on-time deliveries with late deliveries.
# .dropna() is used to handle any missing review scores.
t_stat, p_value = ttest_ind(on_time['review_score'].dropna(), late['review_score'].dropna())
print(f'T-statistic: {t_stat}')
print(f'P-value: {p_value}')

# Interpretation of the results:
# The t-test aims to assess if the mean customer satisfaction (review_score) differs significantly between the two groups (on-time vs. late deliveries).
# - T-statistic: This value measures the size of the difference between the means of the two groups relative to the variation within the groups. A larger absolute T-statistic suggests a greater difference between the group means.
# - P-value: This is the probability of observing a difference as extreme as, or more extreme than, the one calculated, assuming that there is no actual difference between the groups (the null hypothesis).

# As a general rule:
# - If the P-value < 0.05 (a common significance level), we reject the null hypothesis. This means there is a statistically significant difference in customer satisfaction between on-time and late deliveries.
# - If the P-value >= 0.05, we fail to reject the null hypothesis. This suggests there is no statistically significant difference in customer satisfaction between on-time and late deliveries.

# Results:
# A P-value very close to 0.0 observed, indicates a highly significant difference.
# T-statistic (121.496): This value refers to the difference between the average review_score of on-time deliveries and late deliveries,
# relative to the variation within those groups. A large absolute T-statistic (like 121.496) suggests a substantial difference between the mean
# satisfaction scores of the two groups.
# The P-value is 0.0 (extremely small) so we would reject the null hypothesis, concluding that shipping time does affect customer satisfaction.

T-statistic: 121.4963526457059
P-value: 0.0


In [98]:
alpha = 0.05

if p_value < alpha:
  print("Reject H0: There is no difference in customer satisfaction between orders delivered on time and orders delivered late.")
else:
  print("Fail to reject H0: here is no difference in customer satisfaction between orders delivered on time and orders delivered late.")

Reject H0: There is no difference in customer satisfaction between orders delivered on time and orders delivered late.


A/B Testing Process: Divide the dataset into two groups: one group with orders delivered before or on the estimated delivery time and the other with orders delivered after the estimated delivery time. Compute the average satisfaction rating for each group. Conduct a hypothesis test (e.g., t-test) to determine if there is a statistically significant difference in satisfaction ratings between the two groups.

In [97]:
# on_time = df_delivered[df_delivered['order_delivered_customer_date'] <= df_delivered['order_estimated_delivery_date']]
# late = df_delivered[df_delivered['order_delivered_customer_date'] > df_delivered['order_estimated_delivery_date']]



In [93]:
df_ontime = df_delivered[df_delivered['on_time']==True]
df_ontime

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99987  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99988  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99989  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99990  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99991  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99987    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99988    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99989    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99990    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99991    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99987          2017-03-10 11:18:03           2017-03-17 15:08:01   
99988          2018-02-07 23:22:42           2018-02-28 17:37:56   
99989          2017-08-28 20:52:26           2017-09-21 11:24:17   
99990          2018-01-12 15:35:03           2018-01-25 23:32:54   
99991          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date                         review_id  \
0               2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1               2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2               2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3               2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4               2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   
...                             ...                               ...   
99987           2017-03-28 00:00:00  e262b3f92d1ce917aa412a9406cf61a6   
99988           2018-03-02 00:00:00  29bb71b2760d0f876dfa178a76bc4734   
99989           2017-09-27 00:00:00  371579771219f6db2d830d50805977bb   
99990           2018-02-15 00:00:00  8ab6855b9fe9b812cd03a480a25058a1   
99991           2018-04-03 00:00:00  dc9c59b4688062c25758c2be4cafc523   

       review_score review_comment_title  \
0               4.0                  NaN   
1               4.0     Muito boa a loja   
2               5.0                  NaN   
3               5.0                  NaN   
4               5.0                  NaN   
...             ...                  ...   
99987           5.0                  NaN   
99988           4.0                  NaN   
99989           5.0                  NaN   
99990           2.0                  NaN   
99991           5.0                  NaN   

                                  review_comment_message review_cr

In [95]:
df_delivered['review_score_avg_by_delivery_status'] = df_delivered['on_time'].map({True: mean_satisfaction_ontime, False: mean_satisfaction_late})
display(df_delivered.head())

/tmp/ipython-input-2285548278.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  shipping_time_days  avg_shipping_time_days  \
0     2017-10-12 03:43:48                 8.0               12.099537   
1     2018-08-08 18:37:50                13.0               12.099537   
2     2018-08-22 19:07:58                 9.0               12.099537   
3     2017-12-05 19:21:58                13.0               12.099537   
4     2018-02-18 13:02:51                 2.0               12.099537   

   median_shipping_time_days  std_shipping_time_days  on_time  \
0                       10.0                9.547594     True   
1                       10.0                9.547594     True   
2                       10.0                9.547594     True   
3                       10.0                9.547594     True   
4                       10.0                9.547594     True   

   review_score_avg_by_delivery_status  
0                             4.293718  
1                             4.293718  
2                             4.293718  
3                             4.293718  
4                             4.293718

The next step is to calculate and display the mean customer satisfaction (review_score) for both the 'on-time' and 'late' delivery groups. This will directly address the prompt's request to 'Compute the average satisfaction rating for each group'.
Calculate Average Satisfaction
Calculate and display the mean customer satisfaction (review_score) for both the 'on-time' and 'late' delivery groups.


**Reasoning**:
To calculate the mean customer satisfaction for on-time deliveries, I will access the 'review_score' column of the `df_ontime` DataFrame and compute its mean. Then to calculate the mean customer satisfaction for late deliveries, I will access the 'review_score' column of the `late` DataFrame and compute its mean. Finally, I will print both mean values for clear presentation.



In [96]:
mean_satisfaction_ontime = df_ontime['review_score'].mean()
mean_satisfaction_late = late['review_score'].mean()

print(f"Mean customer satisfaction for on-time deliveries: {mean_satisfaction_ontime:.2f}")
print(f"Mean customer satisfaction for late deliveries: {mean_satisfaction_late:.2f}")

Mean customer satisfaction for on-time deliveries: 4.29
Mean customer satisfaction for late deliveries: 2.57


Summary of Hypothesis Test Results

Normality Test (Shapiro-Wilk)
We first performed Shapiro-Wilk tests on the `review_score` distributions for both on-time and late deliveries to check for normality. The results showed:
- On-time deliveries: p-value = `8.060000311605506e-146`
- Late deliveries: p-value = `3.961329385975819e-73`

Both p-values are significantly less than the conventional significance level of 0.05. This indicates that the distributions of `review_score` for both groups (on-time and late deliveries) are not equally distributed. This is in contradiction with a key assumption for parametric tests like the t-test, suggesting that a non-parametric alternative might be more appropriate.

Mann-Whitney U Test
Given the non-normal distribution of our data, the non-parametric Mann-Whitney U test (from cell `X3Kc9EGgiPgn`) was conducted to compare customer satisfaction between on-time and late deliveries. The results are:
- U statistic: `530208489.0`
- p-value: `0.0`

With a p-value of 0.0, which is far below the significance level of 0.05, we cannot validate the null hypothesis (H0) that there is no difference in customer satisfaction between orders delivered on time and orders delivered late. The `alternative='greater'` setting implies that on-time deliveries have significantly higher review scores than late deliveries.

Independent Samples T-test
Although the normality assumption was not met, an independent samples t-test was also performed for comparison. The results were:
- T-statistic: `121.4963526457059`
- P-value: `0.0`

Similar to the Mann-Whitney U test, the t-test also yielded a p-value of 0.0, leading to the rejection of the null hypothesis. The large T-statistic indicates a substantial difference in the mean satisfaction scores. The mean satisfaction for on-time deliveries is `4.29` while for late deliveries it is `2.57`.

Conclusion
Based on the Shapiro-Wilk tests, the `review_score` data for both groups (on-time and late deliveries) is not normally distributed. Therefore, the Mann-Whitney U test is the more appropriate statistical test for comparing customer satisfaction between these two groups.

Both the Mann-Whitney U test and the t-test produced a p-value of 0.0. This overwhelmingly strong evidence leads us to reject the null hypothesis (H0: There is no difference in customer satisfaction between orders delivered on time and orders delivered late). Instead, we accept the alternative hypothesis (H1: Orders delivered on time have higher customer satisfaction compared to orders delivered late).

In conclusion, there is a statistically significant difference in customer satisfaction based on shipping time. Specifically, customers who receive their orders on time report significantly higher satisfaction scores (mean: `4.29`) compared to those whose orders are delivered late (mean: `2.57`). This suggests that timely delivery is a critical factor influencing customer satisfaction.

Visualize Satisfaction Differences

Generate a bar chart comparing the average review scores of on-time versus late deliveries to visually represent the difference in customer satisfaction, including appropriate labels and a legend.

Reasoning:
To prepare for plotting, I will create a new DataFrame that consolidates the average review scores for both on-time and late deliveries into a format suitable for a bar chart. This aligns with the first instruction to create a DataFrame with 'Delivery Status' and 'Average Review Score' columns.



In [90]:
satisfaction_data = pd.DataFrame({
    'Delivery Status': ['On-time', 'Late'],
    'Average Review Score': [mean_satisfaction_ontime, mean_satisfaction_late]
})
satisfaction_data

Delivery Status  Average Review Score
0         On-time              4.293718
1            Late              2.566494

**Reasoning**:
Now that the aggregated data is ready in the `satisfaction_data` DataFrame, I will use `plotly.express` to generate a bar chart to visually compare the average review scores between on-time and late deliveries, adhering to the specified labels and title.



In [91]:
fig = px.bar(satisfaction_data,
    x='Delivery Status',
    y='Average Review Score',
    title='Average Customer Satisfaction by Delivery Status',
    labels={'Delivery Status': 'Delivery Status', 'Average Review Score': 'Average Review Score'})
fig.show()

Summary:

The analysis clearly demonstrates a significant effect of shipping time on customer satisfaction. Orders delivered on time lead to substantially higher customer satisfaction compared to late deliveries.

Data Analysis Key Findings
- The mean customer satisfaction score for on-time deliveries is 4.29, which is significantly higher than the 2.57 recorded for late deliveries.
- Normality tests (Shapiro-Wilk) showed that the `review_score` distributions for both on-time (p-value = 8.06e-146) and late (p-value = 3.96e-73) deliveries are not normally distributed, suggesting that non-parametric tests are more appropriate.
- The Mann-Whitney U test, a non-parametric test, yielded a p-value of 0.0, indicating a statistically significant difference in customer satisfaction, with on-time deliveries resulting in higher satisfaction.
- An independent samples t-test, performed for comparison despite the violated normality assumption, also resulted in a p-value of 0.0 and a T-statistic of 121.50, further confirming a significant difference.
- Both statistical tests reject the null hypothesis, providing strong evidence that on-time deliveries lead to significantly higher customer satisfaction.
 - A bar chart visualization further highlighted the substantial difference in average review scores between on-time and late deliveries.


